src/shims/foreign_items.rs

```diff
@@ -351,13 +356,40 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
     /// Emulates calling a foreign item using its name.
     fn emulate_foreign_item_by_name(
         &mut self,
+        def_id: DefId,
         link_name: Symbol,
         abi: Abi,
         args: &[OpTy<'tcx, Tag>],
         dest: &PlaceTy<'tcx, Tag>,
     ) -> InterpResult<'tcx, EmulateByNameResult<'mir, 'tcx>> {
         let this = self.eval_context_mut();
 
+        let tcx = this.tcx.tcx;
+
+        // First deal with any external C functions in linked .so file
+        // (if any SO file is specified).
+        if this.machine.external_c_so_lib.as_ref().is_some() {
+            if let Some(local_id) = def_id.as_local() {
+                if let Some(extern_c_fct_rep) = ExternalCFuncDeclRep::from_hir_node(
+                    &tcx.hir().get(tcx.hir().local_def_id_to_hir_id(local_id)),
+                    link_name,
+                ) {
+                    // An Ok(false) here means that the function being called was not exported
+                    // by the specified SO file; we should continue and check if it corresponds to
+                    // a provided shim.
+
+                    // TODO ellen! For now, continue to try the shims if there was an error in calling
+                    // the external function.
+                    // As discussed: https://github.com/rust-lang/miri/pull/2363#discussion_r922392879
+                    if let Ok(true) =
+                        this.call_and_add_external_c_fct_to_context(extern_c_fct_rep, dest, args)
```

Comment:
Oh cool, thanks!

Reply:
Hmm, [the solution I have](https://github.com/rust-lang/miri/pull/2363/commits/e7a6466dfb54f43855bf9ddcde5396fa4947873f#diff-316670adf2dc787f3e0c3fef1e129a5f41304a9afa80ca918d2ac4eb4297ffbaR184) ended up being a bit of a hack the `libloading` crate will automatically load system libraries like `libc`. I'm not convinced this is a better solution than what was there before @RalfJung @oli-obk

Reply:
Uh, that is odd. Is this intended behavior of `libloading` to provide symbols that are not exposed by the .so file we are loading?

Reply:
Yeah, on linux [`libloading` is based on `dlsym`](https://docs.rs/libloading/0.7.3/src/libloading/os/unix/mod.rs.html#202) and [`dlsym`](https://linux.die.net/man/3/dlsym) looks through the dependency tree of the library if it can't find the symbol in the library itself. 
I don't really like the idea of restricting the build of the shared library to not depend on `libc`. The main solutions I see to this are:
1. leave in this hack solution
2. go back to checking the shims instead of bailing out if there is any error in calling a function from the library
3. check the shims before checking the library

Reply:
Isn't it possible to have a library that completely hides its dependencies?

> If the symbol is not found, in the specified library or any of the libraries that were automatically loaded by dlopen() when that library was loaded, dlsym() returns NULL.

Wow what a crazy semantics not to have a way to opt-out of that...

Reply:
I'd say we go with the hack (solution 1). What do you think?

Reply:
We do have some kind of precedence here -- `#[no_mangle]` functions declared in Rust can be called via `extern` declarations. And there, we currently *first* try the shim, and only if there is no shim we fall back to these `no_mangle` functions. Additionally we raise an error if a `no_mangle` function is declared for a shim (that happens in `check_abi_and_shim_symbol_clash`).

So I wonder if we should do the same here?
- Have a way to check if a function is actually in that .so file, not one of its dependencies -- looks like you found a way to do that.
- Ensure that there is no overlap between the provided shims and the functions exported from the .so file.

Of course this means adding more shims can break .so file users, which also does not seem great...

Reply:
I think people will actually want the ability to override shims with FFI functions.

Ivwas considering some flags for controlling this, but we can do that when we actually encounter such a situation in practice.

Reply:
> I think people will actually want the ability to override shims with FFI functions.

I agree. But then would they also want to do that with functions exported from Rust? There we have the check because this can be unsound (when you overwrite a function like `malloc` or `memcpy` with one that has the wrong behavior).

Reply:
Considering no one has complained yet, maybe this is not an issue?

For FFI I'd expect this to come up because `malloc` and `free` in C code won't call the miri shims, so something allocated in Rust code can't be freed in FFI code unless we always call the FFI allocator functions

Reply:
Okay, let's document for now that the .so file takes precedence. There should be a warning saying that Miri has its own file descriptor table, so if you overwrite *any* function that works on file descriptors, you have to overwrite *all* of them -- otherwise you are going to be mixing up Miri and host FDs and that's going to be "fun".